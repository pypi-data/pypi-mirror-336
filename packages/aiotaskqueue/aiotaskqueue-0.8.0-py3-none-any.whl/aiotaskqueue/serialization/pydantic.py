from __future__ import annotations

from pydantic import BaseModel

from ._serialization import SerializationBackend, SerializationBackendId


class PydanticSerializer(SerializationBackend[BaseModel]):
    id = SerializationBackendId("pydantic")

    def serializable(self, value: BaseModel) -> bool:
        return isinstance(value, BaseModel)

    def serialize(self, value: BaseModel) -> bytes:
        return value.model_dump_json().encode()

    def deserialize(self, value: bytes, type: type[BaseModel]) -> BaseModel:
        return type.model_validate_json(value)
