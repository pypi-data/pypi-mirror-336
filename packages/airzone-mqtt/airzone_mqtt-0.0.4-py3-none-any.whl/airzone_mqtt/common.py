"""Airzone MQTT Common."""

from __future__ import annotations

from datetime import datetime
from enum import IntEnum
from typing import Any

from .const import TZ_UTC


class AirzoneMode(IntEnum):
    """Airzone MQTT HVAC modes."""

    UNKNOWN = -1

    STOP = 0
    AUTO = 1
    COOLING = 2
    HEATING = 3
    VENTILATION = 4
    DRY = 5
    EMERGENCY_HEAT = 6
    HEAT_AIR = 7
    HEAT_RADIANT = 8
    HEAT_COMBINED = 9
    COOL_AIR = 10
    COOL_RADIANT = 11
    COOL_COMBINED = 12
    BYPASS = 13
    RECOVERY = 14
    REGULATION_TEMP = 15
    PURIFICATION = 16
    FAN_PURIFICATION = 17
    FAN_ENERGY = 18

    @classmethod
    def _missing_(cls, value: Any) -> AirzoneMode:
        return cls.UNKNOWN


class TemperatureUnit(IntEnum):
    """Airzone MQTT temperature units."""

    CELSIUS = 0
    FAHRENHEIT = 1


def get_current_dt() -> datetime:
    """Get current datetime."""
    return datetime.now(tz=TZ_UTC)
