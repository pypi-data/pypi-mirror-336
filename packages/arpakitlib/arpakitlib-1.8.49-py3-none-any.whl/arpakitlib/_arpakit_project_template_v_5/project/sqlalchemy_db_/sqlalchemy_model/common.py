from datetime import datetime
from typing import Any

import sqlalchemy
from sqlalchemy import func
from sqlalchemy.orm import mapped_column, Mapped

from arpakitlib.ar_datetime_util import now_utc_dt
from arpakitlib.ar_sqlalchemy_util import get_string_info_from_declarative_base, BaseDBM
from project.sqlalchemy_db_.util import generate_default_long_id


class SimpleDBM(BaseDBM):
    __abstract__ = True

    id: Mapped[int] = mapped_column(
        sqlalchemy.BIGINT,
        primary_key=True,
        autoincrement=True,
        sort_order=-103,
        nullable=False
    )
    long_id: Mapped[str] = mapped_column(
        sqlalchemy.TEXT,
        insert_default=generate_default_long_id,
        server_default=func.gen_random_uuid(),
        unique=True,
        sort_order=-102,
        nullable=False
    )
    slug: Mapped[str | None] = mapped_column(
        sqlalchemy.TEXT,
        unique=True,
        sort_order=-101,
        nullable=True
    )
    creation_dt: Mapped[datetime] = mapped_column(
        sqlalchemy.TIMESTAMP(timezone=True),
        insert_default=now_utc_dt,
        server_default=func.now(),
        index=True,
        sort_order=-100,
        nullable=False
    )
    extra_data: Mapped[dict[str, Any]] = mapped_column(
        sqlalchemy.JSON,
        index=False,
        nullable=False,
        insert_default={},
        server_default="{}",
        sort_order=1000,
    )

    def __repr__(self) -> str:
        if self.slug is None:
            return f"{self.__class__.__name__.removesuffix('DBM')} (id={self.id})"
        return f"{self.__class__.__name__.removesuffix('DBM')} (id={self.id}, slug={self.slug})"

    @property
    def entity_name(self) -> str:
        return self.__class__.__name__.removesuffix("DBM")

    @property
    def sdp_entity_name(self) -> str:
        return self.entity_name


def get_simple_dbm_class() -> type[SimpleDBM]:
    from project.sqlalchemy_db_.sqlalchemy_model import SimpleDBM
    return SimpleDBM


if __name__ == '__main__':
    print(get_string_info_from_declarative_base(get_simple_dbm_class()))
