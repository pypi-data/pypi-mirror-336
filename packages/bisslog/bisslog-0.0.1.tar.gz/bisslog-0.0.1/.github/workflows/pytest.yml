name: Python Tests with Coverage

on:
  workflow_call:
    inputs:
      coverage_threshold:
        description: "Minimum percentage of coverage required"
        required: false
        default: "80"
        type: string
  workflow_dispatch:
    inputs:
      coverage_threshold:
        description: "Minimum percentage of coverage required"
        required: false
        default: "80"
        type: string

jobs:
  test:
    runs-on: ubuntu-latest

    strategy:
      matrix:
        python-version: ["3.8", "3.9", "3.10", "3.11"]

    steps:
      - name: Checkout Code
        uses: actions/checkout@v4

      - name: Set up Python ${{ matrix.python-version }}
        uses: actions/setup-python@v4
        with:
          python-version: ${{ matrix.python-version }}

      - name: Install Dependencies
        run: |
          python -m pip install --upgrade pip
          if [ -f requirements.txt ]; then pip install -r requirements.txt; fi
          pip install pytest pytest-cov

      - name: Run Tests with Coverage
        run: |
          COV_THRESHOLD=${{ inputs.coverage_threshold }}  # Convertir a variable de shell
          pytest --cov=bisslog --cov-report=term --cov-fail-under=$COV_THRESHOLD tests/
