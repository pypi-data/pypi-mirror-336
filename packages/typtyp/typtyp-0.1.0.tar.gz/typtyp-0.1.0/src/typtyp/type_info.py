import dataclasses


@dataclasses.dataclass(frozen=True)
class TypeInfo:
    name: str
    type: type
    doc: str | None = None

    # Consider `null`s in e.g. dataclasses as `undefined` in TypeScript
    null_is_undefined: bool = False
