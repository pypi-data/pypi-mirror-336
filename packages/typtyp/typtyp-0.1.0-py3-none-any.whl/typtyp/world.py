from __future__ import annotations

import io
from typing import Iterable

from typtyp.type_info import TypeInfo


class _Sentinel:
    pass


NOT_SET = _Sentinel()


class World:
    def __init__(self) -> None:
        self._types_by_name: dict[str, TypeInfo] = {}
        self._types_by_type: dict[type, TypeInfo] = {}

    def add(
        self,
        /,
        typ: type,
        *,
        name: str | None = None,
        doc: str | None | _Sentinel = NOT_SET,
        null_is_undefined: bool = False,
    ) -> TypeInfo:
        if name is None:
            name = typ.__name__
        if isinstance(doc, _Sentinel):
            # You're not a real doc...
            real_doc = getattr(typ, "__doc__", None)
        else:
            real_doc = doc
        assert name  # noqa: S101
        if name in self._types_by_name:
            raise KeyError(f"Type {name} already exists")
        info = TypeInfo(name=name, type=typ, doc=real_doc, null_is_undefined=null_is_undefined)
        self._types_by_name[name] = info
        self._types_by_type[typ] = info
        return info

    def add_many(
        self,
        types: tuple[type, ...] | list[type] | set[type] | dict[str, type],
        *,
        doc: str | None | _Sentinel = NOT_SET,
        null_is_undefined: bool = False,
    ) -> dict[type, TypeInfo]:
        types_it: Iterable[tuple[str | None, type]]
        if isinstance(types, dict):
            types_it = types.items()
        else:
            types_it = ((None, typ) for typ in types)
        ret = {}
        for name, typ in types_it:
            ret[typ] = self.add(typ, name=name, doc=doc, null_is_undefined=null_is_undefined)
        return ret

    def get_name_for_type(self, t: type) -> str:
        return self._types_by_type[t].name

    def __iter__(self):
        return iter(self._types_by_name.values())

    def get_typescript(self) -> str:
        from typtyp.typescript import write_ts

        sio = io.StringIO()
        write_ts(sio, self)
        return sio.getvalue()
