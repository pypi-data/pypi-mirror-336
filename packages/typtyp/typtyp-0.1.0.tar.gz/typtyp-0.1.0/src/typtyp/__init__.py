from typtyp.world import World

__all__ = [
    "World",
]
