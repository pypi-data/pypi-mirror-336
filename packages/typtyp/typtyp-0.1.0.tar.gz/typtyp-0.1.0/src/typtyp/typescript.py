from __future__ import annotations

import collections
import collections.abc
import dataclasses
import datetime
import decimal
import enum
import inspect
import ipaddress
import json
import pathlib
import re
import textwrap
import types
import typing
import uuid
from collections.abc import Iterable
from typing import TYPE_CHECKING, Any, ForwardRef, NamedTuple, TextIO, TypeVar

from typtyp.excs import UnreferrableTypeError
from typtyp.type_info import TypeInfo

if TYPE_CHECKING:
    from typtyp.world import World


RECORD_KEY_TS_TYPE = "string | number | symbol"


def is_pydantic_model(t) -> bool:
    try:
        from pydantic import BaseModel

        return issubclass(t, BaseModel)
    except ImportError:  # pragma: no cover
        return False


class TypeAndKind(NamedTuple):
    type: type
    kind: str | None


@dataclasses.dataclass(frozen=True)
class TypeScriptContext:
    fp: TextIO
    world: World
    null_is_undefined: bool = False
    required_utility_types: dict[str, type] = dataclasses.field(default_factory=dict)

    def sub(self, **replacements):
        return dataclasses.replace(self, **replacements)

    def write(self, s: str) -> None:
        self.fp.write(s)


def map_plain_type_ref(field_type: type, ts_context: TypeScriptContext) -> str:  # noqa: C901, PLR0911, PLR0912
    try:
        # This will handle enums as well â€“ they need to have been registered
        return ts_context.world.get_name_for_type(field_type)
    except KeyError:
        pass

    # Special types

    if isinstance(field_type, ForwardRef):
        return f"unknown /* {field_type.__forward_arg__} */"
    if type(field_type) is TypeVar:
        return f"unknown /* {field_type} */"
    if type(field_type) is type(Ellipsis):
        return "unknown /* ... */"
    if field_type is Any:
        return "unknown /* any */"

    if issubclass(field_type, (bytes, bytearray, memoryview)):
        return f"unknown /* {field_type.__name__} */"

    if issubclass(field_type, bool):
        return "boolean"

    if issubclass(field_type, complex):
        return "[number, number] /* complex */"

    if issubclass(field_type, (pathlib.Path, ipaddress._IPAddressBase, re.Pattern)):
        return f"string /* {field_type.__name__} */"

    if issubclass(field_type, (int, float, decimal.Decimal, datetime.timedelta)) and not issubclass(
        field_type,
        enum.Enum,
    ):
        return "number"

    if issubclass(field_type, uuid.UUID):
        ts_context.required_utility_types["UUID"] = str
        return "UUID"

    if issubclass(field_type, str) and not issubclass(field_type, enum.Enum):
        return "string"

    if issubclass(field_type, type(None)):
        if ts_context.null_is_undefined:
            return "undefined"
        return "null"

    if field_type is datetime.date:
        ts_context.required_utility_types["ISO8601Date"] = str
        return "ISO8601Date"

    if field_type is datetime.time:
        ts_context.required_utility_types["ISO8601Time"] = str
        return "ISO8601Time"

    if field_type is datetime.datetime:
        ts_context.required_utility_types["ISO8601"] = str
        return "ISO8601"

    if issubclass(field_type, collections.Counter):
        return f"Record<{RECORD_KEY_TS_TYPE}, number> /* Counter */"

    if issubclass(field_type, dict):
        comment = f" /* {field_type.__name__} */" if field_type is not dict else ""
        return f"Record<{RECORD_KEY_TS_TYPE}, unknown>{comment}"

    raise UnreferrableTypeError(f"Unable to refer to the type {field_type!r}; if it's a struct, add it to the world")


def to_ts_function_declaration(field_type: type, ts_context: TypeScriptContext) -> str:
    tps = typing.get_args(field_type)
    if len(tps) == 2:
        args_list, retval = tps
        arglist = ", ".join(f"_{i}: {to_ts_type(arg, ts_context)}" for i, arg in enumerate(args_list))
        return f"({arglist}) => {to_ts_type(retval, ts_context)}"
    if len(tps) == 0:
        return "Function"
    raise AssertionError(f"Expected Callable with 0 or 2 types, got {tps}")


def to_ts_type(field_type: type, ts_context: TypeScriptContext) -> str:  # noqa: C901, PLR0911, PLR0912
    if isinstance(field_type, typing.NewType):
        tp = to_ts_type(field_type.__supertype__, ts_context)  # pyright: ignore
        return f"{tp} /* {field_type.__name__} */"

    origin = typing.get_origin(field_type)

    if origin is collections.abc.Callable:
        return to_ts_function_declaration(field_type, ts_context)

    if origin in (typing.Union, types.UnionType):
        return " | ".join(to_ts_type(sub, ts_context) for sub in typing.get_args(field_type))

    if origin is tuple:
        tps = typing.get_args(field_type)
        if tps and tps[-1] is Ellipsis:
            if len(tps) != 2:
                raise AssertionError(f"Expected tuple with one type and Ellipsis, got {tps}")
            return f"[...{to_ts_type(tps[0], ts_context)}[]]"

        return f"[{', '.join(to_ts_type(tp, ts_context) for tp in tps)}]"

    if origin is typing.Literal:
        return " | ".join(json.dumps(arg) for arg in typing.get_args(field_type))

    # TODO: Smells like this could be done better with the ABCs...
    if origin in (
        list,
        collections.deque,
        set,
        frozenset,
        collections.abc.Iterable,
        collections.abc.Iterator,
        collections.abc.Sequence,
    ):
        comment = f" /* {origin.__name__} */" if origin is not list else ""
        tps = typing.get_args(field_type)
        if len(tps) != 1:
            raise AssertionError(f"Expected list with one type, got {tps}")
        inner = to_ts_type(tps[0], ts_context)
        if inner.rstrip("[]").isalnum():
            return f"({inner})[]{comment}"
        return f"Array<{inner}>{comment}"

    if origin is collections.Counter:
        tps = typing.get_args(field_type)
        if len(tps) != 1:
            raise AssertionError(f"Expected Counter with one type, got {tps}")
        return f"Record<{to_ts_type(tps[0], ts_context)}, number>"

    if origin in (
        dict,
        collections.defaultdict,
        collections.abc.Mapping,
        collections.abc.MutableMapping,
        collections.OrderedDict,
    ):
        comment = f" /* {origin.__name__} */" if origin is not dict else ""
        tps = typing.get_args(field_type)
        if len(tps) != 2:
            raise AssertionError(f"Expected dict with two types, got {tps}")
        return f"Record<{to_ts_type(tps[0], ts_context)}, {to_ts_type(tps[1], ts_context)}>{comment}"

    if origin is not None:
        raise NotImplementedError(f"Unknown origin {origin!r} for {field_type!r}")  # pragma: no cover

    if hasattr(field_type, "_fields") and issubclass(field_type, tuple):
        # NamedTuple defined inline? I guess, why not...
        fields: list[str] = field_type._fields  # pyright: ignore
        contents = (f"unknown /* {name} */" for name in fields)
        return f"[{', '.join(contents)}] /* {field_type.__name__} */"

    return map_plain_type_ref(field_type, ts_context)


def get_struct_types(tp) -> list[tuple[str, Any]] | None:
    if dataclasses.is_dataclass(tp):
        return [(f.name, f.type) for f in dataclasses.fields(tp)]

    if typing.is_typeddict(tp):
        # TODO: support __total__
        # TODO: support __required_keys__
        # TODO: support __optional_keys__
        return list(inspect.get_annotations(tp).items())

    try:
        if is_pydantic_model(tp):
            return [(name, f.annotation) for (name, f) in sorted(tp.model_fields.items(), key=lambda item: item[0])]
    except TypeError:  # pragma: no cover
        pass

    return None


def write_structlike(ctx: TypeScriptContext, type_info: TypeInfo, name_and_type: Iterable[tuple[str, type]]) -> None:
    ctx = ctx.sub(null_is_undefined=type_info.null_is_undefined)
    ctx.write(f"interface {type_info.name} {{\n")
    for name, typ in name_and_type:
        ts_type = to_ts_type(typ, ts_context=ctx).strip()
        field_suffix = ""
        if "| undefined" in ts_type:
            ts_type = ts_type.replace("| undefined", "")
            field_suffix = "?"
        ctx.write(f"{name}{field_suffix}: {ts_type}\n")
    ctx.write("}\n")


def write_enum(ctx: TypeScriptContext, type_info: TypeInfo) -> None:
    ctx.write(f"const enum {type_info.name} {{\n")
    for name, value in type_info.type.__members__.items():  # type: ignore
        ctx.write(f"{name} = {json.dumps(value.value)},\n")
    ctx.write("}\n")


def write_type(ctx: TypeScriptContext, type_info: TypeInfo) -> None:
    if type_info.doc:
        ctx.write("/**\n")
        for line in textwrap.dedent(type_info.doc).strip().splitlines():
            ctx.write(f" * {line}\n")
        ctx.write(" */\n")

    if isinstance(type_info.type, type) and issubclass(type_info.type, enum.Enum):
        return write_enum(ctx, type_info)

    if (nt := get_struct_types(type_info.type)) is not None:
        write_structlike(ctx, type_info, nt)
    else:
        ctx.write(f"type {type_info.name} = {to_ts_type(type_info.type, ctx)}\n")


def write_ts(fp: typing.TextIO, world: World) -> None:
    ctx = TypeScriptContext(fp=fp, world=world)
    for type_info in ctx.world:
        write_type(ctx, type_info)
    for name, typ in sorted(ctx.required_utility_types.items()):
        write_type(ctx, TypeInfo(name=name, type=typ))
