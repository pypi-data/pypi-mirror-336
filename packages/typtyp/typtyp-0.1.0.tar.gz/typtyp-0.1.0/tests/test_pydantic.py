from typing import TYPE_CHECKING

import pytest

import typtyp
from tests.common import HairColor
from tests.helpers import check_with_tsc

if TYPE_CHECKING:
    import pydantic
else:
    pydantic = pytest.importorskip("pydantic")


class Head(pydantic.BaseModel):
    size: int
    hair_color: HairColor | None


class Person(pydantic.BaseModel):
    head: Head
    name: str


def test_pydantic(snapshot):
    w = typtyp.World()
    w.add_many((Person, HairColor, Head))
    code = w.get_typescript()
    assert code == snapshot
    assert check_with_tsc(code)
