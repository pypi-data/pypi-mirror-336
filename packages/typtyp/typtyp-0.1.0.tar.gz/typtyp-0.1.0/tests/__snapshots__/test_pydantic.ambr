# serializer version: 1
# name: test_pydantic
  '''
  interface Person {
  head: Head
  name: string
  }
  const enum HairColor {
  RED = "red",
  BLACK = "black",
  PURPLE = "purple",
  BLUE = "blue",
  }
  interface Head {
  hair_color: HairColor | null
  size: number
  }
  
  '''
# ---
